runner/src/runner.rs

```diff
@@ -0,0 +1,275 @@
+                != ChargeResult::Enough
+            {
+                return RunResult::trap_with(
+                    "Not enough gas for initial memory.",
+                    program,
+                    gas_counter.burned(),
+                );
+            };
+        } else if gas_counter
+            .charge(settings.config.load_page_cost * program.get_pages().len() as u64)
+            != ChargeResult::Enough
+        {
+            return RunResult::trap_with(
+                "Not enough gas for loading memory.",
+                program,
+                gas_counter.burned(),
+            );
+        };
+
+        // Creating memory.
+        let memory = env.create_memory(program.static_pages());
+
+        // Charging gas for future growths.
+        if let Some(max_page) = program.get_pages().iter().next_back() {
+            let max_page_num = *max_page.0;
+            let mem_size = memory.size();
+            if max_page_num >= mem_size {
+                let amount =
+                    settings.config.mem_grow_cost * ((max_page_num - mem_size).raw() as u64 + 1);
+
+                if gas_counter.charge(amount) != ChargeResult::Enough {
+                    return RunResult::trap_with(
+                        "Not enough gas for grow memory size.",
+                        program,
+                        gas_counter.burned(),
+                    );
+                }
+            } else {
+                assert!(max_page_num.raw() == mem_size.raw() - 1);
+            }
+        }
+
+        // Getting allocations.
+        let allocations: BTreeSet<PageNumber> = program
+            .get_pages()
+            .iter()
+            .map(|(page_num, _)| *page_num)
+            .collect();
+
+        // Creating memory context.
+        let memory_context = MemoryContext::new(
+            program.id(),
+            memory.clone(),
+            allocations,
+            program.static_pages().into(),
+            settings.config.max_pages,
+        );
+
+        // Creating message context.
+        let message_context = MessageContext::new(
+            message.clone(),
+            BlakeMessageIdGenerator {
+                program_id: program.id(),
+                nonce: program.message_nonce(),
+            },
+        );
+
+        // Creating externalities.
+        let ext = Ext {
+            gas_counter,
+            memory_context,
+            message_context,
+            block_info: settings.block_info,
+            config: settings.config,
+            error_explanation: None,
+            waited: false,
+        };
+
+        // Running backend.
+        let (res, mut ext) = env.setup_and_run(
+            ext,
+            instrumented_code,
+            program.get_pages(),
+            &*memory,
+            settings.entry.into(),
+        );
+
+        // Parsing outcome.
+        let outcome = if let Err(e) = res {
+            let explanation = ext.error_explanation.take().unwrap_or("N/A");
```

Comment:
why not leave it `Option<&'static str>` instead of generating magic string?

Reply:
Option added back